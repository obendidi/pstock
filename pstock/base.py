from __future__ import annotations

import typing as tp
from abc import ABC, abstractmethod
from datetime import datetime

import pandas as pd
import pendulum
from pydantic import BaseModel as _BaseModel
from pydantic import PrivateAttr


class BaseModel(_BaseModel):
    _created_at: datetime = PrivateAttr(default_factory=pendulum.now)

    @property
    def created_at(self) -> datetime:
        return self._created_at


class BaseModelDf(BaseModel, ABC):

    _df: tp.Optional[pd.DataFrame] = PrivateAttr(default=None)

    @abstractmethod
    def gen_df(self) -> pd.DataFrame:
        ...

    @property
    def df(self) -> pd.DataFrame:
        if self._df is None:
            self._df = self.gen_df()
        return self._df


T = tp.TypeVar("T", bound=BaseModel)


class BaseModelSequence(tp.Generic[T], BaseModelDf):

    __root__: tp.Sequence[T]

    @tp.overload
    def __getitem__(self, index: int) -> T:
        """Get single item from __root__ by idx."""

    @tp.overload
    def __getitem__(self, index: slice) -> tp.Sequence[T]:
        """Get slice of items from __root__ by idx."""

    def __getitem__(self, index):
        return self.__root__[index]

    def __len__(self) -> int:
        return len(self.__root__)

    def __iter__(self) -> tp.Iterator[T]:  # type: ignore
        return iter(self.__root__)

    def gen_df(self) -> pd.DataFrame:
        return pd.DataFrame.from_dict(self.dict().get("__root__"), orient="columns")


U = tp.TypeVar("U", bound=BaseModelSequence)


class BaseModelMapping(tp.Generic[U], BaseModelDf):

    __root__: tp.Mapping[str, U]

    def __getitem__(self, index: str) -> U:
        return self.__root__[index]

    def __len__(self) -> int:
        return len(self.__root__)

    def __iter__(self) -> tp.Iterator[str]:  # type: ignore
        return iter(self.__root__)

    def gen_df(self) -> pd.DataFrame:
        keys, dfs = zip(*[(key, value.df) for key, value in self.__root__.items()])
        return pd.concat(dfs, axis=1, keys=keys)
