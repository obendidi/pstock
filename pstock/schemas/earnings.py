import typing as tp

import pandas as pd
from pydantic import BaseModel

from pstock.schemas.base import BaseDataFrameModel


class Earning(BaseModel):
    quarter: str
    estimate: float
    actual: float
    revenue: float
    earnings: float


class Earnings(BaseDataFrameModel):
    __root__: tp.List[Earning]

    @property
    def df(self) -> pd.DataFrame:
        if self._df is None:
            self._df = self._convert_to_df(index_column="quarter")
        return self._df
