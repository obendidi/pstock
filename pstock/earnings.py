import typing as tp

import numpy as np
import pandas as pd
from pydantic import validator

from pstock.base import BaseModel, BaseModelSequence
from pstock.quote import QuoteSummary
from pstock.utils.quote import get_earnings_data_from_quote


class Earning(BaseModel):
    quarter: str
    estimate: float
    actual: float
    status: tp.Literal[None, "Beat", "Missed"] = None
    revenue: float
    earnings: float

    @validator("status", always=True)
    def set_status(
        cls, value: tp.Any, values: tp.Dict[str, tp.Any]
    ) -> tp.Literal[None, "Beat", "Missed"]:
        if value is not None:
            return value
        estimate = values.get("estimate")
        actual = values.get("actual")
        if actual is None or np.isnan(actual) or estimate is None or np.isnan(estimate):
            return None
        return "Beat" if actual >= estimate else "Missed"


class Earnings(BaseModelSequence[Earning], QuoteSummary):
    __root__: tp.List[Earning]

    def gen_df(self) -> pd.DataFrame:
        df = super().gen_df()
        if not df.empty:
            df = df.set_index("quarter").sort_index(key=pd.to_datetime)
        return df

    @validator("__root__")
    def sort_earnings(cls, value: tp.List[Earning]) -> tp.List[Earning]:
        if not value:
            return value
        return sorted(value, key=lambda earning: pd.to_datetime(earning.quarter))

    @classmethod
    def process_quote(cls, quote: tp.Dict[str, tp.Any]) -> tp.Dict[str, tp.Any]:
        return {"__root__": get_earnings_data_from_quote(quote)}
