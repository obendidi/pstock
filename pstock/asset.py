from __future__ import annotations

import typing as tp

import asyncer
import httpx
import numpy as np
import pandas as pd
from pydantic import Field, validator

from pstock.base import BaseModelSequence
from pstock.earnings import Earnings
from pstock.income_statement import IncomeStatements
from pstock.quote import QuoteSummary
from pstock.trend import Trends
from pstock.utils.quote import get_asset_data_from_quote
from pstock.utils.utils import httpx_client_manager


class Asset(QuoteSummary):
    symbol: str
    name: str
    asset_type: tp.Literal[
        "EQUITY", "CURRENCY", "CRYPTOCURRENCY", "ETF", "FUTURE", "INDEX"
    ]
    currency: str
    latest_price: float = np.nan
    sector: tp.Optional[str]
    industry: tp.Optional[str]
    earnings: Earnings = Field(repr=False)
    trends: Trends = Field(repr=False)
    income_statement: tp.Optional[IncomeStatements] = Field(repr=False)

    @validator("symbol")
    def symbol_upper(cls, symbol: str) -> str:
        return symbol.upper()

    @classmethod
    def process_quote(cls, quote: tp.Dict[str, tp.Any]) -> tp.Dict[str, tp.Any]:
        data = get_asset_data_from_quote(quote)
        earnings = Earnings.process_quote(quote)
        trends = Trends.process_quote(quote)
        return {**data, "earnings": earnings, "trends": trends}

    @classmethod
    def process_financials_quote(
        cls, financials_quote: tp.Dict[str, tp.Any]
    ) -> tp.Dict[str, tp.Any]:
        income_statement = IncomeStatements.process_financials_quote(financials_quote)
        return {"income_statement": income_statement}


class Assets(BaseModelSequence[Asset]):
    __root__: tp.List[Asset]

    def gen_df(self) -> pd.DataFrame:
        df = super().gen_df()
        df["earnings"] = df["earnings"].apply(lambda v: None if len(v) == 0 else v)
        df["trends"] = df["trends"].apply(lambda v: None if len(v) == 0 else v)
        return df.set_index("symbol").sort_index().dropna(axis=1, how="all")

    @classmethod
    async def get(
        cls,
        symbols: tp.List[str],
        *,
        client: tp.Optional[httpx.AsyncClient] = None,
    ):
        async with httpx_client_manager(client=client) as _client:
            async with asyncer.create_task_group() as tg:
                soon_values = [
                    tg.soonify(Asset.get)(symbol, client=_client) for symbol in symbols
                ]
        return cls.parse_obj([soon.value for soon in soon_values])
