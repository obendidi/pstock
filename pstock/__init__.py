from pstock.asset import Asset, Assets
from pstock.bar import Bars, BarsMulti
from pstock.earnings import Earnings
from pstock.income_statement import IncomeStatements
from pstock.news import News
from pstock.trend import Trends
from pstock.utils.utils import rdm_user_agent_value
