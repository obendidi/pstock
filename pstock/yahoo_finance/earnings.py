import typing as tp

import httpx
import numpy as np

from pstock.schemas import Earning, Earnings
from pstock.yahoo_finance.quote import get_quote_summary

__all__ = "get_earnings"


def _parse_quarterly_earnings(
    quarterly_earnings: tp.List[tp.Dict[str, tp.Any]],
    quarterly_financial_chart: tp.List[tp.Dict[str, tp.Any]],
) -> tp.List[Earning]:
    date_to_earnings = {
        e.get("date", ""): {
            "actual": e.get("actual", {}).get("raw", np.nan),
            "estimate": e.get("estimate", {}).get("raw", np.nan),
        }
        for e in quarterly_earnings
        if "date" in e
    }
    date_to_fin_chart = {
        c.get("date", ""): {
            "revenue": c.get("revenue", {}).get("raw", np.nan),
            "earnings": c.get("earnings", {}).get("raw", np.nan),
        }
        for c in quarterly_financial_chart
        if "date" in c
    }
    all_dates = set(list(date_to_earnings.keys()) + list(date_to_fin_chart.keys()))

    return [
        Earning(
            quarter=quarter,
            actual=date_to_earnings.get(quarter, {}).get("actual", np.nan),
            estimate=date_to_earnings.get(quarter, {}).get("estimate", np.nan),
            revenue=date_to_fin_chart.get(quarter, {}).get("revenue", np.nan),
            earnings=date_to_fin_chart.get(quarter, {}).get("earnings", np.nan),
        )
        for quarter in all_dates
    ]


def _parse_next_earnings(earnings_chart: tp.Dict[str, tp.Any]) -> Earning:
    return Earning(
        quarter=(
            f"{earnings_chart.get('currentQuarterEstimateDate', '')}"
            f"{earnings_chart.get('currentQuarterEstimateYear', '')}"
        ),
        estimate=earnings_chart.get("currentQuarterEstimate", {}).get("raw", np.nan),
        actual=np.nan,
        revenue=np.nan,
        earnings=np.nan,
    )


def _get_next_earnings_date(earnings_chart: tp.Dict[str, tp.Any]) -> tp.Optional[int]:
    e_dates = earnings_chart.get("earningsDate", [])
    if not e_dates:
        return None
    return e_dates[0].get("raw")


def _parse_earnings_from_quote_summary(
    quote_summary: tp.Dict[str, tp.Any]
) -> tp.Tuple[Earnings, tp.Optional[int]]:
    earnings = quote_summary.get("earnings")
    if earnings is None or not isinstance(earnings, dict):
        return Earnings.parse_obj([]), None

    earnings_chart = earnings.get("earningsChart", {})
    quarterly_earnings = earnings_chart.get("quarterly", [])
    quarterly_financial_chart = earnings.get("financialsChart", {}).get("quarterly", [])

    parsed_list = _parse_quarterly_earnings(
        quarterly_earnings, quarterly_financial_chart
    ) + [_parse_next_earnings(earnings_chart)]
    nex_earnings_date = _get_next_earnings_date(earnings_chart)
    return Earnings.parse_obj(parsed_list), nex_earnings_date


async def get_earnings(
    symbol: str, client: tp.Optional[httpx.AsyncClient] = None
) -> Earnings:
    """Get [Earnings][pstock.schemas.Earnings] data from yahoo-finance.

    [Earnings][pstock.schemas.Earnings] are parsed from the `quote_summary` generated by
    [get_quote_summary][pstock.yahoo_finance.get_quote_summary].

    Is provided symbol has no earnings (crypto/ETF) and empty
    [Earnings][pstock.schemas.Earnings] is returned.

    Args:
        symbol (str): A stock symbol availlable in yahoo-finance
        client (tp.Optional[httpx.AsyncClient], optional): Defaults to None.

    Returns:
        [pstock.schemas.Earnings][]
    """
    quote_summary = await get_quote_summary(symbol, client=client)
    earnings, _ = _parse_earnings_from_quote_summary(quote_summary)
    return earnings


if __name__ == "__main__":
    import logging

    import asyncer

    from pstock.core.log import setup_logging

    setup_logging(level="INFO")
    logger = logging.getLogger()

    async def _worker(symbol: str, client: httpx.AsyncClient) -> None:
        asset = await get_earnings(symbol, client=client)
        logger.info(symbol)
        logger.info(asset.df)

    async def _main(symbols):
        async with httpx.AsyncClient() as client:
            async with asyncer.create_task_group() as tg:
                for symbol in symbols:
                    tg.soonify(_worker)(symbol, client=client)

    asyncer.runnify(_main)(
        [
            "TSLA",
            "AAPL",
            "GOOG",
            "AMZN",
            "AMD",
            "GME",
            "SPCE",
            "^QQQ",
            "ETH-USD",
            "BTC-EUR",
        ]
    )
