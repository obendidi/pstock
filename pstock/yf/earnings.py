import typing as tp

import httpx

from pstock.earnings import Earnings

from .quote import get_yf_quote_summary_content_async, get_yf_quote_summary_content_sync


def get_earnings_sync(
    symbol: str, client: tp.Optional[httpx.Client] = None
) -> Earnings:
    """Get [Earnings][pstock.schemas.Earnings] data from yahoo-finance.

    [Earnings][pstock.schemas.Earnings] are parsed from the `quote_summary` generated by
    [get_quote_summary][pstock.yahoo_finance.get_quote_summary].

    Is provided symbol has no earnings (crypto/ETF) and empty
    [Earnings][pstock.schemas.Earnings] is returned.

    Args:
        symbol (str): A stock symbol availlable in yahoo-finance
        client (tp.Optional[httpx.Client], optional): Defaults to None.

    Returns:
        [pstock.schemas.Earnings][]
    """
    content = get_yf_quote_summary_content_sync(symbol, client=client)
    return Earnings.from_yf(content=content)


async def get_earnings_async(
    symbol: str, client: tp.Optional[httpx.AsyncClient] = None
) -> Earnings:
    """Get [Earnings][pstock.schemas.Earnings] data from yahoo-finance.

    [Earnings][pstock.schemas.Earnings] are parsed from the `quote_summary` generated by
    [get_quote_summary][pstock.yahoo_finance.get_quote_summary].

    Is provided symbol has no earnings (crypto/ETF) and empty
    [Earnings][pstock.schemas.Earnings] is returned.

    Args:
        symbol (str): A stock symbol availlable in yahoo-finance
        client (tp.Optional[httpx.AsyncClient], optional): Defaults to None.

    Returns:
        [pstock.schemas.Earnings][]
    """
    content = await get_yf_quote_summary_content_async(symbol, client=client)
    return Earnings.from_yf(content=content)
