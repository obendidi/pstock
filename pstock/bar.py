from __future__ import annotations

import json
import logging
import typing as tp
from datetime import datetime, timedelta
from urllib.parse import urlencode

import asyncer
import httpx
import pandas as pd
import pendulum
from pydantic import validate_arguments

from pstock.base import BaseModel, BaseModelMapping, BaseModelSequence
from pstock.types import ReadableResponse, Timestamp
from pstock.utils import httpx_client_manager, parse_datetime, parse_duration
from pstock.yahoo_finance.chart import get_ohlc_from_chart

logger = logging.getLogger(__name__)

IntervalParam = tp.Literal[
    "1m", "2m", "5m", "15m", "30m", "1h", "1d", "5d", "1mo", "3mo"
]
PeriodParam = tp.Literal[
    "1d", "5d", "1mo", "3mo", "6mo", "1y", "2y", "5y", "10y", "ytd", "max"
]
EventParam = tp.Literal["div", "split", "div,splits"]


def _get_lowest_valid_interval(
    *, period: tp.Optional[PeriodParam], start: tp.Optional[Timestamp]
) -> IntervalParam:
    if period == "max":
        return "3mo"

    if period is not None:
        delta = parse_duration(period)
    elif start is not None:
        delta = pendulum.now() - parse_datetime(start)
    else:
        raise ValueError(f"period={period}, start={start}")

    if delta <= pendulum.duration(days=7):
        return "1m"
    elif delta <= pendulum.duration(days=60):
        return "2m"
    elif delta <= pendulum.duration(days=730):
        return "1h"
    else:
        return "1d"


class Bar(BaseModel):
    date: datetime
    open: float
    high: float
    low: float
    close: float
    adj_close: float
    volume: float
    interval: timedelta


class _BarMixin:
    @classmethod
    def base_uri(cls, symbol: str) -> str:
        return f"https://query2.finance.yahoo.com/v8/finance/chart/{symbol.upper()}"

    @classmethod
    @validate_arguments
    def params(
        cls,
        interval: tp.Optional[IntervalParam] = None,
        period: tp.Optional[PeriodParam] = None,
        start: tp.Optional[Timestamp] = None,
        end: tp.Optional[Timestamp] = None,
        events: EventParam = "div,splits",
        include_prepost: bool = False,
    ) -> tp.Dict[str, tp.Any]:
        if period is None and start is None:
            period = "max"

        if end is None and start is not None:
            end = pendulum.now().int_timestamp

        if interval is None:
            interval = _get_lowest_valid_interval(period=period, start=start)
            print(f"Setting interval to: {interval}")

        _params: tp.Dict[str, tp.Any] = {
            "interval": interval,
            "events": events,
            "includePrePost": include_prepost,
        }
        if period is not None:
            _params["range"] = period
        if start is not None:
            _params["period1"] = start
        if end is not None:
            _params["period2"] = end

        return _params

    @classmethod
    def uri(
        cls,
        symbol: str,
        *,
        interval: tp.Optional[IntervalParam] = None,
        period: tp.Optional[PeriodParam] = None,
        start: tp.Optional[Timestamp] = None,
        end: tp.Optional[Timestamp] = None,
        events: EventParam = "div,splits",
        include_prepost: bool = False,
    ) -> str:
        params = cls.params(
            interval=interval,
            period=period,
            start=start,
            end=end,
            events=events,
            include_prepost=include_prepost,
        )
        return f"{cls.base_uri(symbol)}?{urlencode(params)}"


class Bars(BaseModelSequence[Bar], _BarMixin):
    __root__: tp.List[Bar]

    def _gen_df(self) -> pd.DataFrame:
        df = super()._gen_df()
        if not df.empty:
            df = df.dropna(
                how="all",
                subset=["open", "high", "low", "close", "adj_close", "volume"],
            )
            if df["interval"][0] >= timedelta(days=1):
                df["date"] = pd.to_datetime(df["date"]).dt.date
            df = df.set_index("date").sort_index()
        return df

    @classmethod
    def load(
        cls,
        *,
        response: tp.Union[ReadableResponse, str, bytes, dict],
    ) -> Bars:
        if isinstance(response, dict):
            data = response
        elif isinstance(response, (str, bytes)):
            data = json.loads(response)
        else:
            data = json.loads(response.read())

        return cls.parse_obj(get_ohlc_from_chart(data))

    @classmethod
    async def get(
        cls,
        symbol: str,
        *,
        interval: tp.Optional[IntervalParam] = None,
        period: tp.Optional[PeriodParam] = None,
        start: tp.Optional[Timestamp] = None,
        end: tp.Optional[Timestamp] = None,
        events: EventParam = "div,splits",
        include_prepost: bool = False,
        client: tp.Optional[httpx.AsyncClient] = None,
    ):
        url = cls.base_uri(symbol)
        params = cls.params(
            interval=interval,
            period=period,
            start=start,
            end=end,
            events=events,
            include_prepost=include_prepost,
        )
        async with httpx_client_manager(client=client) as _client:
            response = await _client.get(url, params=params)

        response.raise_for_status()
        return cls.load(response=response)


class BarsMulti(BaseModelMapping[Bars], _BarMixin):
    __root__: tp.Dict[str, Bars]

    def _gen_df(self) -> pd.DataFrame:
        df = super()._gen_df()
        return df.sort_index()

    @classmethod
    async def get(
        cls,
        symbols: tp.List[str],
        *,
        interval: tp.Optional[IntervalParam] = None,
        period: tp.Optional[PeriodParam] = None,
        start: tp.Optional[Timestamp] = None,
        end: tp.Optional[Timestamp] = None,
        events: EventParam = "div,splits",
        include_prepost: bool = False,
        client: tp.Optional[httpx.AsyncClient] = None,
    ):
        async with httpx_client_manager(client=client) as _client:
            async with asyncer.create_task_group() as tg:
                soon_values = [
                    tg.soonify(Bars.get)(
                        symbol,
                        interval=interval,
                        period=period,
                        start=start,
                        end=end,
                        include_prepost=include_prepost,
                        events=events,
                        client=_client,
                    )
                    for symbol in symbols
                ]
        data = {
            symbol: soon_value.value for symbol, soon_value in zip(symbols, soon_values)
        }
        return cls.parse_obj(data)
