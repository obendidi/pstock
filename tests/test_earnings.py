import httpx
import pandas as pd

from pstock.earnings import Earnings


def test_earnings(main_quote_response: httpx.Response, snapshot):
    earnings = Earnings.load(response=main_quote_response)
    assert earnings.dict() == snapshot(name="pydantic-model")
    assert isinstance(earnings.df, pd.DataFrame)
    assert earnings.df.to_dict() == snapshot(name="pandas-dataframe")
