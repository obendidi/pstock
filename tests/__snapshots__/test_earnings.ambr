# name: test_earnings[cryptocurrency_quote][pandas-dataframe]
  <class 'dict'> {
  }
---
# name: test_earnings[cryptocurrency_quote][pydantic-model]
  <class 'dict'> {
    '__root__': <class 'list'> [
    ],
  }
---
# name: test_earnings[currency_quote][pandas-dataframe]
  <class 'dict'> {
  }
---
# name: test_earnings[currency_quote][pydantic-model]
  <class 'dict'> {
    '__root__': <class 'list'> [
    ],
  }
---
# name: test_earnings[equity_quote][pandas-dataframe]
  <class 'dict'> {
    'actual': <class 'dict'> {
      '1Q2021': 0.93,
      '1Q2022': nan,
      '2Q2021': 1.45,
      '3Q2021': 1.86,
      '4Q2021': 2.54,
    },
    'earnings': <class 'dict'> {
      '1Q2021': 438000000.0,
      '1Q2022': nan,
      '2Q2021': 1142000000.0,
      '3Q2021': 1618000000.0,
      '4Q2021': 2321000000.0,
    },
    'estimate': <class 'dict'> {
      '1Q2021': 0.79,
      '1Q2022': 2.24,
      '2Q2021': 0.98,
      '3Q2021': 1.59,
      '4Q2021': 2.36,
    },
    'revenue': <class 'dict'> {
      '1Q2021': 10389000000.0,
      '1Q2022': nan,
      '2Q2021': 11958000000.0,
      '3Q2021': 13757000000.0,
      '4Q2021': 17719000000.0,
    },
    'status': <class 'dict'> {
      '1Q2021': 'Beat',
      '1Q2022': None,
      '2Q2021': 'Beat',
      '3Q2021': 'Beat',
      '4Q2021': 'Beat',
    },
  }
---
# name: test_earnings[equity_quote][pydantic-model]
  <class 'dict'> {
    '__root__': <class 'list'> [
      <class 'dict'> {
        'actual': 0.93,
        'earnings': 438000000.0,
        'estimate': 0.79,
        'quarter': '1Q2021',
        'revenue': 10389000000.0,
        'status': 'Beat',
      },
      <class 'dict'> {
        'actual': 1.45,
        'earnings': 1142000000.0,
        'estimate': 0.98,
        'quarter': '2Q2021',
        'revenue': 11958000000.0,
        'status': 'Beat',
      },
      <class 'dict'> {
        'actual': 1.86,
        'earnings': 1618000000.0,
        'estimate': 1.59,
        'quarter': '3Q2021',
        'revenue': 13757000000.0,
        'status': 'Beat',
      },
      <class 'dict'> {
        'actual': 2.54,
        'earnings': 2321000000.0,
        'estimate': 2.36,
        'quarter': '4Q2021',
        'revenue': 17719000000.0,
        'status': 'Beat',
      },
      <class 'dict'> {
        'actual': nan,
        'earnings': nan,
        'estimate': 2.24,
        'quarter': '1Q2022',
        'revenue': nan,
        'status': None,
      },
    ],
  }
---
# name: test_earnings[etf_quote][pandas-dataframe]
  <class 'dict'> {
  }
---
# name: test_earnings[etf_quote][pydantic-model]
  <class 'dict'> {
    '__root__': <class 'list'> [
    ],
  }
---
# name: test_earnings[future_quote][pandas-dataframe]
  <class 'dict'> {
  }
---
# name: test_earnings[future_quote][pydantic-model]
  <class 'dict'> {
    '__root__': <class 'list'> [
    ],
  }
---
# name: test_earnings[index_quote][pandas-dataframe]
  <class 'dict'> {
  }
---
# name: test_earnings[index_quote][pydantic-model]
  <class 'dict'> {
    '__root__': <class 'list'> [
    ],
  }
---
